### PR TITLE
Add PIC/PIE tests and fix some bugs + some improvements to the test harness

### LINKED ISSUE #22052: Add PIE tests for all relevant targets
`std.os.linux.pie` has been ported to all(?) supported architectures. We should add some module tests that build and run with PIE.

### PR DESCRIPTION
**Pull request by alexrp**
Closes #22052.

### CODE REVIEW

File: test/cases/spirv_mergable_pointers.zig [hunk 1]
```diff
@@ -14,3 +14,4 @@ export fn a() void {
 // output_mode=Obj
 // backend=stage2
 // target=spirv64-vulkan
+// emit_bin=false
```
Comment (alexrp) [inline test/cases/spirv_mergable_pointers.zig:17]: FYI @Snektron 1d1ca8485454a4b14f20376aff532bfe7c959155

File: test/cases/inherit_want_safety.zig [hunk 2]
```diff
@@ -35,3 +35,4 @@ pub export fn entry() usize {
 
 // compile
 // output_mode=Obj
+// emit_bin=false
```
Comment (alexrp) [inline test/cases/inherit_want_safety.zig:38]: FYI @andrewrk 5beb5f20d282fd0db8715899ba6a5d6630d91bb4 you'll probably want to revert this one once you finish your current wasm linker work.

File: test/src/Cases.zig [hunk 3]
```diff
@@ -705,6 +725,10 @@ pub fn lowerToBuildSteps(
 
         switch (update.case) {
             .Compile => {
+                // Force the binary to be emitted if requested.
+                if (case.emit_bin) {
+                    _ = artifact.getEmittedBin();
+                }
```
Comment (alexrp) [inline test/src/Cases.zig:731]: This feels very hacky; definitely open to better ideas.

File: test/cases/spirv_mergable_pointers.zig (hunk 1 above)
Comment (Snektron) [reply test/cases/spirv_mergable_pointers.zig:17]: Should be fine

File: test/src/Cases.zig (hunk 3 above)
Comment (alexrp) [reply test/src/Cases.zig:731]: Well, it seems `std.Build` itself uses this pattern in some places, so at least there's precedent... though it still seems like there ought to be a more user-friendly way to tell the build system that codegen is required.

**Pull request merged**